### PR TITLE
More control for merging stores together

### PR DESCRIPTION
**Pull request by mpiannucci**
in favor of #154 ..

### CODE REVIEW

File: icechunk-python/examples/dask_write.py [hunk 1]
```diff
@@ -188,13 +188,15 @@ def update(args: argparse.Namespace) -> None:
...
-    worker_stores = client.gather(map_result)
+    worker_changes = client.gather(map_result)
 
     print("Starting distributed commit")
     # we can use the current store as the commit coordinator, because it doesn't have any pending changes,
     # all changes come from the tasks, Icechunk doesn't care about where the changes come from, the only
     # important thing is to not count changes twice
-    commit_res = store.distributed_commit("distributed commit", [ws.change_set_bytes() for ws in worker_stores])
+    for changes in worker_changes:
+        store.merge(changes)
```
Comment (dcherian) [inline icechunk-python/examples/dask_write.py:198]: Thoughts on `store.merge(stores)` instead? Is there a benefit to merging the changeset bytes?

Comment (mpiannucci) [reply icechunk-python/examples/dask_write.py:198]: I am ok with it, I was just thinking of use cases where the full stores aren't serialized but it's kind of a non issue

Comment (dcherian) [reply icechunk-python/examples/dask_write.py:198]: I agree with it being a non-issue

File: icechunk/src/repository.rs [hunk 2]
```diff
@@ -922,20 +907,18 @@ async fn get_existing_node<'a>(
     }
 }
 
-async fn distributed_flush<I: IntoIterator<Item = ChangeSet>>(
+async fn flush(
     storage: &(dyn Storage + Send + Sync),
-    change_sets: I,
+    change_set: &mut ChangeSet,
```
Comment (dcherian) [inline icechunk/src/repository.rs:912]: I don't understand why this needs a mutable ref.

**Review by dcherian — APPROVED** [review]
Just that one comment about the mutable ref for my understanding.

Otherwise this seems OK to me!

File: icechunk/src/repository.rs [hunk 3]
```diff
@@ -635,26 +635,15 @@ impl Repository {
...
+    /// Discard all uncommitted changes and return them as a `ChangeSet`
+    pub fn discard_changes(&mut self) -> ChangeSet {
+        std::mem::take(&mut self.change_set)
+    }
 
-        self.snapshot_id = new_snapshot_id.clone();
-        self.change_set = ChangeSet::default();
-        Ok(new_snapshot_id)
+    /// Merge a set of `ChangeSet`s into the repository without committing them
+    pub async fn merge(&mut self, changes: ChangeSet) -> RepositoryResult<()> {
```
Comment (paraseba) [inline icechunk/src/repository.rs:644]: I don't think this can fail? let's turn it into `-> ()`

**Review by paraseba — APPROVED** [review]

File: icechunk/src/repository.rs (hunk 2 above)
Comment (mpiannucci) [reply icechunk/src/repository.rs]: It was a relic from the distributed flush! Good catch!

**Pull request merged**